### DISCUSSION TITLE
What is the proper way to shut down BullMQ?

### DISCUSSION THREAD

**Discussion by nullndr** [taskforcesh/bullmq#2051] [Q&A]
I'm building a huge application with a lot of workers and queues and I want to correctly handle node signals.

Currently I have the following:

```typescript
process.on("SIGINT", async (signal) => {
  logger(`I received a ${signal} signal.`);
  await closeWorkers();
  await disconnectDb();
  process.exit(130);
});
```

Where the `closeWorkers()` function is the following and follows the [docs](https://docs.bullmq.io/guide/going-to-production#gracefully-shut-down-workers):

```typescript
const closeWorkers = async () => {
  await Promise.all([
    foo.worker.close(),
    bar.worker.close(),
    fooBar.worker.close(),
  ]);
};
```

Now, should I also call the `close()` method on the queue instance? If yes, I should call it after the workers, right?

And what is the best approach to handle this when the redis instance is down? I mean when the `retryStategy()` still fails.

**Comment by manast** [comment]
The reason that it is important to close a worker when shutting down a service is that it may be processing jobs and by gracefully closing it reduces the chance of having jobs that stall. This is not a problem for the Queue instance, however, it should always be a good practice to close all your connections and Queues are connected to Redis so I would also close them.

**Reply by nullndr** [reply]
Yeah but they should be closed after the workers, right?

**Reply by manast** [reply]
Why?
